### PR TITLE
chore(deps): mark typescript as optional

### PR DESCRIPTION
**Pull request by kripod**
## PR Checklist

- [x] Addresses an existing open issue: related to #370
- [x] Steps in [Contributing](https://github.com/ts-essentials/ts-essentials/blob/master/CONTRIBUTING.md) were taken

## Overview

This marks `typescript` as optional so that JS-only apps may import code from libraries dependent upon ts-essentials without having to install TypeScript.

The version constraint for consumers using `typescript` would still apply.

### CODE REVIEW

**Review by Beraliv — COMMENTED** [review]
Thanks for PR @kripod!

The only thing that I'd like you to add is a new row in [TypeScript dependency table](https://github.com/ts-essentials/ts-essentials#typescript-dependency-table)

Next ts-essentials release will be v9.4.0 so you can add:

| `ts-essentials` | `typescript` / type of dependency |
| `^9.4.0` | `^4.1.0` / [peer optional](#370) |

Including the link to the issue – #370

**Comment by kripod** [comment]
I just did that, thank you 🙌

**Comment by kripod** [comment]
I just did what you asked for, thank you 🙌

**Review by Beraliv — APPROVED** [review]
Thank you for the help, let me test it in the evening and I will merge it after that

**Review by Beraliv — CHANGES_REQUESTED** [review]
Hey again 👋

Apologies for requesting changes, but I've spotted you've missed changelog in the PR as it directly affects the end user experience.

Could you please add `changelog` by running `yarn changeset`? You can find recommendations in here – https://github.com/ts-essentials/ts-essentials/blob/master/CONTRIBUTING.md

247454755-90222a98-d3e1-4a95-bca7-86638ae191d6


Once it's done, I'm happy to approve it again

Also please let me know what projects use ts-essentials as a dependency so I can test it locally

**Comment by kripod** [comment]
Hello,

Thanks for your feedback. I’ve just added a changeset as you’ve asked.

Unfortunately, I can’t provide a library which uses ts-essentials as a dependency, as the one I’m working on is private. Please [refer to the npm docs](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#peerdependenciesmeta) to see how `peerDependenciesMeta` works.

**Comment by kripod** [comment]
@Beraliv Hello again,

Please let me know if there’s anything I can help with to get this merged. Take your time with the review, I just wanted to know if I did everything required.

Thank you in advance 🙌

**Comment by Beraliv** [comment]
> @Beraliv Hello again,
> 
> Please let me know if there’s anything I can help with to get this merged. Take your time with the review, I just wanted to know if I did everything required.
> 
> Thank you in advance 🙌

Hey!

Apologies, I was very busy last months. Thank you again for your contribution!

I will test it this week and will let you know if there's anything left to merge this

**Review by Beraliv — APPROVED** [review]

**Comment by kripod** [comment]
Hello @Beraliv,

I appreciate that you’ve taken the time to merge this PR. May I ask if a new npm release might be planned soon? I’d be happy to try the changes in action.

Thank you in advance 🙏

**Comment by Beraliv** [comment]
> Hello @Beraliv,
> 
> 
> 
> I appreciate that you’ve taken the time to merge this PR. May I ask if a new npm release might be planned soon? I’d be happy to try the changes in action.
> 
> 
> 
> Thank you in advance 🙏

Hey!

Apologies for the late release, it's done!

**Comment by QuentinJanuel** [comment]
Could you please clarify why is typescript listed as an optional peer dependency instead of a dev dependency?
Thanks!

**Comment by Beraliv** [comment]
@QuentinJanuel it's in the description of this PR, please have a look at "Overview":

> This marks typescript as optional so that JS-only apps may import code from libraries dependent upon ts-essentials without having to install TypeScript

**Comment by QuentinJanuel** [comment]
Oh sorry I had read the Overview but didn't understand what it implied to use optional peer dependency.
So if I understand correctly, you want to avoid js only users to have to install typescript even as a dev dependency only. The issue is that for us we need to make sure to ignore optionals (which is not a default) to not have the whole library in production.
So it's a tradeoff, it would be great if there were something like optional perr+dev dependencies

**Comment by Beraliv** [comment]
> Oh sorry I had read the Overview but didn't understand why it implied to use optional peer dependency.
> 
> 
> 
> So if I understand correctly, you want to avoid JS-only users to have to install TS even as a dev-dependency only.
> 
> The issue is that for us we need to make sure to ignore optionals (which is not a default) to not have the whole library in production.
> 
> So it's a tradeoff... it would be great if there were something like optional peer+dev dependencies 😕

I see, what's the impact for you right now?

Would it be possible to workaround it by manually adding the typescript dev dependency? Could you please show the project or minimum reproduction so I am on the same page with you?

**Pull request merged**
